[#21] 상품 등록 메서드 리팩토링

rbsks opened the pull request:
### 변경사항
<!-- 이 PR에서 어떤점들이 변경되었는지 기술해주세요. 가급적이면 as-is, to-be를 활용해서 작성해주세요.  -->
**AS-IS**
- validationCreateStoreProductRequest 메서드에서 모든 데이터의 유효성을 검증을 하고 있어 가독성 저하
- createStoreProduct에서 상품, 이미지, 옵션 생성하는 코드가 메서드로 분리되지 않아 가독성 저하

**TO-BE**
- validationCreateStoreProductRequest 내부에서 더 작은 단위로 분리된 유효성 검증 메서드를 사용하여 가독성을 증가 시키고 각 유효성 메서드가 단일 책임 원칙을 지키도록 수정
- 상품, 이미지, 옵션 엔티티 생성에 대한 메서드를 분리하여 가독성 향상
- 의미를 가지는 메서드 이름으로 추출하였기 때문에 추후에 코드를 다시 보더라도 각 메서드들이 어떤 역할을 하고있는지 쉽게 알 수 있음

### 테스트
<!-- 본 변경사항이 테스트가 되었는지 기술해주세요 --> 
- [x] 테스트 코드
- [x] API 테스트

rbsks committed:
Refactor: Exception refactor

rbsks committed:
Refactor: createStoreProduct method refactoring
- 유효성 검증 메서드를 작은 단위로 분리하여 메서드도 단일 책임 원칙을 지키도록 수정
- 상품, 이미지, 옵션 엔티티 생성 메서드 추출

f-lab-kyle commented on src/main/java/com/stylelab/file/exception/FileException.java:
```diff
@@ -10,7 +10,7 @@ public class FileException extends ServiceException {
 
     public FileException(CommonError serviceError) {
-        super(serviceError);
+        super(serviceError, serviceError.getMessage());
```
👍

f-lab-kyle commented on src/main/java/com/stylelab/product/application/StoreProductFacade.java:
```diff
@@ -30,187 +37,255 @@ public class StoreProductFacade {
+        return product;
+    }
+
+    private void addProductImages(CreateStoreProductRequestVo createStoreProductRequestVo, Product product) {
```
Product를 파라미터로 받는 이유가 productImages에 add해주기 위해서라면 
이 메소드의 역할은 add할 productImages를 반환하는 것으로 제한하고 (책임 분리)

그걸 받아서 add 하는게 더 좋지 않을까 하는 생각도 되네요. 이를 다시 코드로 표현해보면

```java
product.addProductImages(getProductImages(createStoreProductRequestVo));
```

가 될것 같아요.

rbsks replied:
파라미터도 제한을 하면 더 확실하게 책임의 분리가 되겠네요 ㅎㅎ 감사합니다!

f-lab-kyle commented on src/main/java/com/stylelab/product/application/StoreProductFacade.java:
```diff
@@ -30,187 +37,255 @@ public class StoreProductFacade {
-        }
-        product.addProductImages(productImages);
+        Product product = createProduct(storeId, createStoreProductRequestVo.productRequest());
+        addProductImages(createStoreProductRequestVo, product);
```
아래 메소드 선언부에 리뷰 드리겠습니다~

f-lab-kyle commented on src/main/java/com/stylelab/product/application/StoreProductFacade.java:
```diff
@@ -30,187 +37,255 @@ public class StoreProductFacade {
+        addProductImages(createStoreProductRequestVo, product);
 
-        boolean useOption = productRequest.useOption();
+        boolean useOption = createStoreProductRequestVo.productRequest().useOption();
```
https://www.baeldung.com/java-demeter-law

rbsks replied:
디미터의 법칙은 처음 들어보네요!! 참고하겠습니다!
감사합니다~

f-lab-kyle commented on src/main/java/com/stylelab/product/application/StoreProductFacade.java:
```diff
@@ -30,187 +37,255 @@ public class StoreProductFacade {
-            }
-
-            product.additionalProductOption1(productOption1s);
+            addProductOptions(createStoreProductRequestVo, product);
```
여기도 addProductImages 리뷰와 비슷한 맥락으로 검토해주세요.

f-lab-kyle commented on src/main/java/com/stylelab/product/application/StoreProductFacade.java:
```diff
@@ -30,187 +37,255 @@ public class StoreProductFacade {
         }
 
-        CreateStoreProductRequestVo.EntryMain entryMain = createStoreProductRequestVo.entryMain();
+        validationProductMainImages(createStoreProductRequestVo.entryMain());
+        validationProductSubImages(createStoreProductRequestVo.entrySubs());
+        validationProductDescriptionImages(createStoreProductRequestVo.descriptions());
+        validationProductInformation(createStoreProductRequestVo.productRequest());
+        validationProductOptions(createStoreProductRequestVo);
```
오버로딩으로 해도 괜찮을것 같구요.

rbsks replied:
오버로딩을 사용하게 되면 메서드의 이름이 같아져서 메서드가 무슨 역할을 하는지 잘 모를 거 같은데 이런 경우는 파라미터로 메서드가 어떤 역할을 하는지 판단할 수 있기 때문에 상관없을까요??

f-lab-kyle commented on src/main/java/com/stylelab/product/application/StoreProductFacade.java:
```diff
@@ -30,187 +37,255 @@ public class StoreProductFacade {
             throw new ProductException(
                     ProductError.EXCEED_MAX_IMAGE_COUNT,
-                    String.format(ProductError.EXCEED_MAX_IMAGE_COUNT.getMessage(), ImageType.PRODUCT_ENTRY_SUB, ImageType.PRODUCT_ENTRY_SUB.getMaxImageCount())
+                    String.format(
+                            ProductError.EXCEED_MAX_IMAGE_COUNT.getMessage(),
+                            ImageType.PRODUCT_ENTRY_SUB,
+                            ImageType.PRODUCT_ENTRY_SUB.getMaxImageCount()
+                    )
```
이 메소드도 반복적으로 보여서 별도의 유틸 클래스나 메소드로 추출해서 사용할수 있을것 같네요.

f-lab-kyle commented on src/main/java/com/stylelab/product/application/StoreProductFacade.java:
```diff
@@ -30,187 +37,255 @@ public class StoreProductFacade {
-            }
+            int productQuantity = createStoreProductRequestVo.productRequest().quantity();
+            validationNotUseProductOption(productQuantity, minimumQuantity);
         } else {
```
위 if문으로 들어가면 return하면 되니 else가 필요 없어질것 같아요.

rbsks replied:
그렇네요 ㅎㅎ 수정하였습니다!

rbsks committed:
Refactor: code refactor
- ApiResponse record로 변경
- ApiResponse 변경으로 인한 ServiceExceptionHandler 코드 수정
- CreateStoreProductRequestVo 디미터 법칙 적용
- StoreProductFacade createStoreProduct refactoring
- server port 변경
